pkg/docker/config/config.go

```diff
@@ -170,41 +222,70 @@ func GetCredentials(sys *types.SystemContext, registry string) (types.DockerAuth
 // it exists only to allow testing it with an artificial home directory.
 func getCredentialsWithHomeDir(sys *types.SystemContext, registry, homeDir string) (types.DockerAuthConfig, error) {
 	if sys != nil && sys.DockerAuthConfig != nil {
-		logrus.Debug("Returning credentials from DockerAuthConfig")
+		logrus.Debugf("Returning credentials for %s from DockerAuthConfig", registry)
 		return *sys.DockerAuthConfig, nil
 	}
 
-	if enableKeyring {
-		username, password, err := getAuthFromKernelKeyring(registry)
-		if err == nil {
-			logrus.Debug("returning credentials from kernel keyring")
-			return types.DockerAuthConfig{
-				Username: username,
-				Password: password,
-			}, nil
+	// Anonymous function to query credentials from auth files.
+	getCredentialsFromAuthFiles := func() (types.DockerAuthConfig, error) {
+		for _, path := range getAuthFilePaths(sys, homeDir) {
+			authConfig, err := findAuthentication(registry, path.path, path.legacyFormat)
+			if err != nil {
+				return types.DockerAuthConfig{}, err
+			}
+
+			if (authConfig.Username != "" && authConfig.Password != "") || authConfig.IdentityToken != "" {
+				return authConfig, nil
+			}
 		}
+		return types.DockerAuthConfig{}, nil
 	}
 
-	for _, path := range getAuthFilePaths(sys, homeDir) {
-		authConfig, err := findAuthentication(registry, path.path, path.legacyFormat)
+	helpers, err := sysregistriesv2.FindCredentialHelpers(sys, registry)
+	if err != nil {
+		return types.DockerAuthConfig{}, err
+	}
+	// If the registry doesn't set credential helpers explicitly, use the
+	// global defaults.
+	if len(helpers) == 0 {
+		helpers, err = sysregistriesv2.GlobalCredentialHelpers(sys)
 		if err != nil {
-			logrus.Debugf("Credentials not found")
 			return types.DockerAuthConfig{}, err
 		}
+	}
```

Comment:
At a first glance this looks like something that should be a single call, not something a caller could have the option of getting wrong. (OTOH one caller is already different.)